haskell-api.md

```diff
@@ -0,0 +1,236 @@
+
+## Status Codes
+
+- 201 for creation
+- 204 for no content (i.e. our `Empty` responses)
+- 400s MUST follow our `ValidationError` machinery
+
+## Pagination
+
+- All list-returning routes of modest size should be paginated via `Yesod.Page`
+
+## Modules Structure
+
+- Fully expand all parents to start
+
+  This means we won't break Frontend when the tree expands over time
+
+- Fully name-space all parents and routes (see example)
+
+  **Caveats**:
+
+  - There's no need to `V3`-prefix everything
+  - Things can "start over" when there's the "single below a list" situation,
+    e.g. `TeacherP x $ TeacherR` instead of `TeachersP $ TeachersTeacherP x $
+    TeachersTeacherR`, which is a bit buffalo-buffalo
+  - The "start over" caveat does not apply to path naming (i.e. the above
+    example will still have the module `Teachers/Teacher.hs`)
+
+- Name modules to directly match parents and routes, which implies one Handler
+  module per route (see example)
+
+  **Caveats**:
+
+  - In reality, we need to prevent collisions between versions through
+    additional prefixing of route constructors. We ignore this in this guide.
+
+Example:
+
+```
+/3 V3P:                         --> Handlers/V3
+  /students StudentsP:          --> Handlers/V3/Students
+    / StudentsR GET             --> Handlers/V3/Students.hs
+    /#StudentId StudentP:       --> Handlers/V3/Students/Student
+      / StudentR GET            --> Handlers/V3/Students/Student.hs
+      /courses StudentCoursesP: --> Handlers/V3/Students/Student/Courses
+        / StudentCoursesR GET   --> Handlers/V3/Students/Student/Courses.hs
+```
+
+## Module Structure
+
+- Prefer in-module request/response types
+
+  - Request type naming: `{Resource}{Method}Body`
+  - Response type naming: `{Resource}{Method}`
+  - NOTE: list responses are `[{Resource}{Method}]`, not `{Resources}{Method}`
+
+  ```hs
+  -- Good
+  data TeacherGet
+
+  getTeachersR :: Handler [TeacherGet]
+  getTeachersR = undefined
+
+  data TeacherPost
+  data TeacherPostBody
+
+  postTeacherR :: Handler TeacherPost
+  postTeacherR = do
+    body <- requireJsonBody @TeacherPostBody
+
+  -- Bad
+  data Teachers
+
+  getTeachersR :: Handler Teachers
+  getTeachersR = undefined
+
+  data PostTeacher
+
+  postTeacherR :: Handler (Entity Teacher)
+  postTeacherR = do
+    body <- requireJsonBody @PostTeacher
+  ```
+
+- Prefer in-module data-access functions, until they require sharing
+- Order your module as:
+  - Request/response type(s)
+  - Handler that uses it
+  - Repeat if more than one handler
+  - Internal functions
+- Document your endpoints
+
+See `/3/teachers` as a good example.
+
+## Testing
+
+Do not export your request/response types for use in tests. Instead, re-build
+JSON `Value`s with `object` This ensures you don't have a (de)serialization bug
+that passes the tests because it's used in both places. It also makes it clear
+how to move to `aeson-lens`-based assertions (which are more robust) when that
+makes sense.
```

Comment:
After fiddling with `aeson-lens` this morning, I'm less enthusiastic about this. I find `aesonQQ` plus a reasonably loose predicate a lot more readable.

Reply:
FWIW, I had the same experience, and I know it's not convincing, but after using it in anger over three features I'm absolutely in love. It's an annoying journey, but on the other side has been bliss for me.

I like it because it's so flexible that I can make all sorts of specific assertions that clearly show my intent in the test -- in a consistent way for any assertion. I don't have to do things differently if I want whole-body vs targeted, it's all the same thing to match on a sub-doc, order of keys, attribute, etc.  To me that's more readable on a macro level, not just in the `aesonQQ`-vs-`object` comparison.

That said, I don't feel strongly about pushing against `aesonQQ` -- it's a subjective "me" thing about how I prefer to write my tests. This guide doesn't preclude `aesonQQ` at all, so I can just remove this point too.